### PR TITLE
Feature: Option to disable activate-on-release behaviour of toolbar dropdown buttons.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

A common problem for new users is they don't realise the main toolbar buttons bring up a dropdown menu as it disappears straight after clicking.

Therefore they cannot find how to change railtypes nor find a few more features probably.

## Description

This change introduces an option to change this behaviour (by default) so that clicking and releasing simply opens the dropdown menu.

This is probably game-breaking for existing users but if the default value maintains existing behaviour then it won't benefit new users at all.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/lang/english.txt [hunk 1]
```diff
@@ -1568,6 +1568,9 @@ STR_CONFIG_SETTING_SCROLLMODE_LMB                               :Move map with L
 STR_CONFIG_SETTING_SMOOTH_SCROLLING                             :Smooth viewport scrolling: {STRING2}
 STR_CONFIG_SETTING_SMOOTH_SCROLLING_HELPTEXT                    :Control how the main view scrolls to a specific position when clicking on the smallmap or when issuing a command to scroll to a specific object on the map. If enabled, the viewport scrolls smoothly, if disabled it jumps directly to the targeted spot
 
+STR_CONFIG_SETTING_TOOLBAR_DROPDOWN_CLOSE                       :Toolbar dropdowns activate on release: {STRING2}
+STR_CONFIG_SETTING_TOOLBAR_DROPDOWN_CLOSE_HELPTEXT              :Control whether tthe main toolbar dropdowns activate their default function when released, or if they require an additional selection.
```
Comment (TrueBrain) [inline src/lang/english.txt]: tthe? :D Drop the t please :) Just one, not both.

Comment (TrueBrain) [inline src/lang/english.txt]: This sentence is really hard to read, and even after reading it 4 times, I am not exactly sure what it does :P Maybe dumb it down a bit?

**Review by TrueBrain — COMMENTED** [review]
I like the idea of this PR :)

Does it still mean that if you click it once, it selects the default? Or do you always have to click twice?

**Comment by PeterN** [comment]
You can click-drag-release to select an option with only one click, but simply clicking on the button will just leave the dropdown open, and require another click.

**Comment by TrueBrain** [comment]
> You can click-drag-release to select an option with only one click, but simply clicking on the button will just leave the dropdown open, and require another click.

Okay, current players will find that annoying; as would I. But for new players this will be so much better. So I am wondering if we can make that a bit easier on both parties .. but I still like the idea of this PR a lot :)

**Comment by 2TallTyler** [comment]
I wonder if we should introduce a framework to do what the survey opt-in did, but to ask users to choose their preference when a new setting is introduced. Something like:

```
We've added a new setting. Here's what it does. Would you to try the new default or use the current/legacy behavior?

[use new default] [keep current setting]
```

It might be annoying, but might also cut down on "where did my signals go?" comments. 😃 And we already sort of do this with new engine previews. 😛

**Comment by TrueBrain** [comment]
The other suggestion discussed on Discord the other day, was to use the fact the configuration file is versioned. So if there is an existing config, keep the old setting. And if there is none, use the new.

I like both ideas; one is just more work than the other :P In general, a "What's new" dialog could also be good :)

**Comment by Kuhnovic** [comment]
I like the idea, it's definitely going to be helpful for new players. Keep in mind that we probably need to write a whole bunch of tooltips for these menu items since you can now actually hover over them.

> You can click-drag-release to select an option with only one click, but simply clicking on the button will just leave the dropdown open, and require another click.

Reading this I don't think it's going to be that annoying to players that are used to the old behavior, since it sounds like the old way of mousepress-drag-mouserelease still works. The only thing that would be different to us oldtimers is if we misclicked on the main toolbar button and suddenly the menu would stay, but one click would solve that.

Idea 1: hitting the close-all-windows-hotkey should probably close these dropdown menu's too.

Idea 2: maybe the dropdown should close the moment the user moves the mouse cursor outside of it. Not sure if that would feel natural or not, but it might be worth a try.

**Comment by EmperorJake** [comment]
While the old behaviour was always completely natural to me, I'm surprised at how many new players are confused by it so I would support this change being enabled by default. 
As I said I would use this new behaviour even as a veteran player, because selecting a railtype from the bottom of a long list is very difficult when using a trackpad, especially if the list is long enough to have a scroll bar.

Also +1 on the delete hotkey closing any open dropdowns.

**Comment by Andrew350** [comment]
So I messed around with this on the online preview for about 20 minutes (awesome feature by the way) and had some thoughts.

Overall I found this very intuitive even as a veteran player, mostly because just about every other game in existence for the last 25 years or so has done drop-down/pop-out menus like this. It did trip up my muscle memory at first, but it was actually very easy to adapt - for the most part. I still find it jarring for things where I'm used to doing a one-click action, for example saving a game or auto-selecting the most used railtype (which by the way, this makes that setting basically useless), because it takes one more click than before in those cases. 

There are just certain buttons I have engrained into my head that only need one quick click to do the action I want and it is tough to break that habit 😛 But for everything else where the usual click-and-drag method was necessary I actually find this a welcome change and am 1000% certain it is better for new players. I'm not sure if there's some way of doing a sort of hybrid system (perhaps something like Kuhnovic's suggestion) to try to preserve the best of both worlds, but if not I still think this PR is a better default option.

Whether the aforementioned inconvenience of losing the use of those one click actions will be welcomed by other seasoned players is another story...but I am at least willing to give it a chance, and those who don't can always change it back 🙂

Also, the delete key does indeed close the drop-downs which is nice, however clicking on the same menu button again does not which I feel is unintuitive, but maybe that's just an oversight. Also thinking if the "right click to close windows" setting is on you should be able to close dropdowns on right click as well, if that's possible.

**Comment by EmperorJake** [comment]
I've given this a test now, and I'd say a change that absolutely needs to be made is for clicking the button again to make the menu disappear. This behaviour would be consistent with most other programs/OS that use dropdowns like this, and so it's what I'm used to doing. And if the dropdown has only one item it shouldn't need a dropdown at all, but instead behave like a simple button like it did previously.

But that's all I can complain about, it's a good change and it makes navigating a long list of railtypes more convenient on either a mouse or trackpad. 

Also I'd argue that most-used railtype doesn't become useless, as I make heavy use of the A for autorail hotkey.

**Comment by PeterN** [comment]
Yeah, not disappearing when pressing the button again is a bug, looking into it :-)

**Comment by 2TallTyler** [comment]
See also: #11338.

**Comment by mdias** [comment]
This PR was just brought to my attention as it appears I've made basically a duplicate of it (#11338).

It seems that we've mostly implemented it the same way, but I managed to fix the "click button again for the dropdown to disappear". Please look into the `window.cpp` change on my PR for that.

If you decide to incorporate that change please let me know and I'll close my PR.

**Comment by 2TallTyler** [comment]
Both @PeterN and @mdias have implemented this feature as separate PRs. At this point I think it's fair game for either of you (or anyone else) to borrow code from the other and update their PR with one perfect solution, that we can get into the upcoming 14.0 release. 😄

**Comment by mdias** [comment]
Since @PeterN's PR is considerably older than mine, I think it's fair for his to go in.
I also don't mind if he incorporates my fix to the button behaving properly when clicked again mentioned by [a previous comment](https://github.com/OpenTTD/OpenTTD/pull/10938#issuecomment-1590375501).

**Comment by nielsmh** [comment]
This has been requested on the forums also: https://www.tt-forums.net/viewtopic.php?t=91878

**Comment by AnthonyDiGirolamo** [comment]
New player here, would 100% love this. I'm on a laptop with a touchscreen (GPD Win Max 2) and tapping to open a toolbar menu doesn't work since it immediately disappears again. The Android port seems to do what I'd prefer by default.

**Comment by mdias** [comment]
I'm not sure what needs doing at this point to merge it.
If there's something depending on me, please let me know.

File: src/lang/english.txt [hunk 2]
```diff
@@ -1690,6 +1690,9 @@ STR_CONFIG_SETTING_SCROLLMODE_LMB                               :Move map with L
 STR_CONFIG_SETTING_SMOOTH_SCROLLING                             :Smooth viewport scrolling: {STRING2}
 STR_CONFIG_SETTING_SMOOTH_SCROLLING_HELPTEXT                    :Control how the main view scrolls to a specific position when clicking on the smallmap or when issuing a command to scroll to a specific object on the map. If enabled, the viewport scrolls smoothly, if disabled it jumps directly to the targeted spot
 
+STR_CONFIG_SETTING_TOOLBAR_DROPDOWN_AUTOSELECT                  :Toolbar dropdowns automatically activate: {STRING2}
```
Comment (2TallTyler) [inline src/lang/english.txt]: This is a hard thing to explain, so what about a different approach of not trying to fully explain it? Something like `Quick toolbar dropdowns` or even `Legacy toolbar dropdowns?

File: src/lang/english.txt [hunk 3]
```diff
@@ -1690,6 +1690,9 @@ STR_CONFIG_SETTING_SCROLLMODE_LMB                               :Move map with L
...
+STR_CONFIG_SETTING_TOOLBAR_DROPDOWN_AUTOSELECT_HELPTEXT         :Control whether the main toolbar dropdowns activate their default function when the mouse button is released. If not enabled then the dropdown menu will stay open so a selection can be made.
```
Comment (2TallTyler) [inline src/lang/english.txt]:
```suggestion
STR_CONFIG_SETTING_TOOLBAR_DROPDOWN_AUTOSELECT_HELPTEXT         :Enable legacy main toolbar behavior. If enabled, clicking on a main toolbar button will activate its default function while holding the mouse button will open the dropdown, selecting whichever options is hovered over when released. If disabled, clicking on a main toolbar button will open the dropdown menu without needing to hold the mouse button.
```

Comment (PeterN) [reply src/lang/english.txt]: I think "legacy" is a bit strong, the current way is not being deprecated.

Comment (2TallTyler) [reply src/lang/english.txt]: Traditional?

Comment (PeterN) [reply src/lang/english.txt]: IMHO this text is too wordy, and is wrong. We can omit unnecessary details but if we include them it should be correct. (The dropdown menus are always displayed when clicking, they just close again after a brief time.)

**Review by 2TallTyler — APPROVED** [review]
